model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java

```diff
@@ -795,38 +837,59 @@ public CredentialModel getStoredCredentialByNameAndType(RealmModel realm, UserMo
+        List<MapUserCredentialEntity> newList = credentialEntityMap == null ? new LinkedList<>()
+                : credentialEntityMap.stream()
+                        .sorted(MapUserCredentialEntity.ORDER_BY_PRIORITY)
+                        .collect(Collectors.toList());
+
+        // 2 - Find indexes of our and newPrevious credential
+        int ourCredentialIndex = -1;
+        int newPreviousCredentialIndex = -1;
+        MapUserCredentialEntity ourCredential = null;
+        int i = 0;
+        for (MapUserCredentialEntity credential : newList) {
+            if (id.equals(credential.getId())) {
+                ourCredentialIndex = i;
+                ourCredential = credential;
+            } else if(newPreviousCredentialId != null && newPreviousCredentialId.equals(credential.getId())) {
+                newPreviousCredentialIndex = i;
             }
+            i++;
         }
 
-        // Find current index of credential (id) which will be moved
-        int currentPositionOfId = userEntity.getCredentialIndex(id);
-        if (currentPositionOfId == -1) {
-            LOG.warnf("Credential with id: [%s] for user: [%s] not found", id, userId);
+        if (ourCredentialIndex == -1) {
+            LOG.warnf("Not found credential with id [%s] of user [%s]", id, user.getUsername());
             return false;
         }
 
-        // If id is before newPreviousCredentialId in priority list, it will be moved to position -1
-        if (currentPositionOfId < newPreviousCredentialIdIndex) {
-            newPreviousCredentialIdIndex -= 1;
+        if (newPreviousCredentialId != null && newPreviousCredentialIndex == -1) {
+            LOG.warnf("Can't move up credential with id [%s] of user [%s]", id, user.getUsername());
+            return false;
         }
 
-        // Move credential to desired index
-        userEntity.moveCredential(currentPositionOfId, newPreviousCredentialIdIndex + 1);
+        // 3 - Compute index where we move our credential
+        int toMoveIndex = newPreviousCredentialId==null ? 0 : newPreviousCredentialIndex + 1;
+
+        // 4 - Insert our credential to new position, remove it from the old position
+        newList.add(toMoveIndex, ourCredential);
+        int indexToRemove = toMoveIndex < ourCredentialIndex ? ourCredentialIndex + 1 : ourCredentialIndex;
+        newList.remove(indexToRemove);
+
+        // 5 - newList contains credentials in requested order now. Iterate through whole list and change priorities accordingly.
+        int expectedPriority = 0;
+        for (MapUserCredentialEntity credential : newList) {
+            expectedPriority += PRIORITY_DIFFERENCE;
+            if (credential.getPriority() != expectedPriority) {
+                credential.setPriority(expectedPriority);
+
+                LOG.tracef("Priority of credential [%s] of user [%s] changed to [%d]", credential.getId(), user.getUsername(), expectedPriority);
+            }
+        }
```

Comment:
Nice refactor.
I believe this method belongs mostly, if not completely, to the entity where its logic can be optimized for a particular storage.